keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,625 @@
+when the above batch workload has an autoscale semantic.
+
+Some third-party frameworks have implemented success policy.
+
+- [Kubeflow Training Operator](https://www.kubeflow.org/docs/components/training)
+- [Flux Operator](https://flux-framework.org/flux-operator/)
+- [JobSet](https://github.com/kubernetes-sigs/jobset/)
+
+### Goals
+
+- allow to mark a job as a `Succeeded` according to a declared policy.
+
+### Non-Goals
+
+- Change the existing behavior of `NonIndexed` mode.
+- Change the existing behavior of `Indexed` mode without a SuccessPolicy.
+
+## Proposal
+
+We propose new policies under which a job can be declared as succeeded.
+Those policies can be modeled in the following:
+
+1. An indexed job completes if a set of [x, y, z...] indexes are successful.
+2. An indexed job completes if x or x% of indexes are successful.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a machine-learning researcher, I run an indexed job which leaders are launched
+as index=0,1,2, and workers are launched as index=3+.
+I want to care about only the leaders when the result of job is evaluated.
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  successPolicy:
+    onSucceededPods:
+      indexes: "0-2"
+      whenTargetAchieved: terminate
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: job-container
+        image: job-image
+        command: ["./sample"]
+```
+
+#### Story 2
+
+As a machine-learning researcher, I run an indexed job with an autoscale semantic
+set as a minimum number is 5 and a maximum number is 20.
+I want to mark the job as successful when at least 5 indexes succeeded, even if 
+all indexes don't succeed.
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  successPolicy:
+    onSucceededPods:
+      count: 5
+      whenTargetAchieved: terminate
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: job-container
+        image: job-image
+        command: ["./sample"]
+```
+
+#### Story 3
+
+As a machine-learning researcher from the [Story 2](#story-2), I want to mark the job
+as successful if 50% of all indexes succeed because I can not know how many indexes 
+will have in advance.
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  successPolicy:
+    onSucceededPods:
+      count: 50%
+      whenTargetAchieved: terminate
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+        - name: job-container
+          image: job-image
+          command: ["./sample"]
+```
+
+#### Story 4
+
+As a simulation researcher/engineer for fluid dynamics/biochemistry.
+I want to mark the job as successful and continue running the rest of the pods if some indexes succeed
```

Comment:
I think that your concerns could be resolved by the new condition type `Succeeded`.
@kannon92 WDYT?

Reply:
Yea though conditions are not that granular. My main concern is say you submit  a job with 10,000 pods and you say are okay if 50% of them finish. What happens if a user wants to know what the other 50% did? Since I think most people that submits jobs would want them to actually finish..

Condition can at least say that successPolicy was successful but there are running jobs. UX is still a bit lackluster.

But I think we can't expect conditions to do that. So I'm fine with resolving this.

I guess in this case maybe kube isn't the best thing to use as you probably want a dedicated workflow engine to help with that case..

Reply:
> if 50% of them finish. What happens if a user wants to know what the other 50% did? Since I think most people that submits jobs would want them to actually finish..

@kannon92 Yes, you're right, but I think that the user can find the result of leaving the pod in the status field since even if the `Succeeded` condition is added to the job, the job keeps tracking the leaving pods and updating the job status:

https://github.com/kubernetes/kubernetes/blob/a9e4f5b7862e84c4152eabe2e960f3f6fb9a4867/pkg/apis/batch/types.go#L464-L485